Keep full guarantees in availability assignments state (rho)

zdave-parity opened the pull request:
Full guarantees are needed for auditing:

- To determine the guarantors to try directly fetching the bundle from.
- In the case of a dispute, the guarantor signatures are needed to construct a disputes extrinsic.

It is convenient to have the full guarantees available in the state at the point where auditing begins. Without this, it would be necessary for validator nodes to track them separately from the point of reporting.

Note that `\mathbb{G}` is now used for the set of guarantees. Inner/guest PVM state uses `\mathbb{I}` instead.

Other minor fixes included in this PR:

- Fix order of set definitions. Intended to be alphabetical.
- Fix misnamed macros. \reportspostassurances and \reportspostguarantees were the wrong way around.
- Improve some of the wording around timing out of reports pending availability. Availability assignments are now proactively cleared when the timeout elapses.

zdave-parity committed:
Remove \accinput and use \mathbb{I} for inner/guest PVM state rather than \mathbb{G}

This frees up \mathbb{G} to be used for guarantees.

\accinput, which was \mathbb{I}, isn't really needed.

zdave-parity committed:
Fix misnamed macros

\reportspostassurances and \reportspostguarantees were the wrong way
around.

Additionally, improve some of the wording around timing out of reports
pending availability. Availability assignments are now proactively
cleared when the timeout elapses.

zdave-parity committed:
Fix order of set definitions

Intended to be alphabetical.

zdave-parity committed:
Keep full guarantees in availability assignments state (rho)

Full guarantees are needed for auditing:

- To determine the guarantors to try directly fetching the bundle from.
- In the case of a dispute, the guarantor signatures are needed to
  construct a disputes extrinsic.

It is convenient to have the full guarantees available in the state at
the point where auditing begins. Without this, it would be necessary for
validator nodes to track them separately from the point of reporting.

alxmirap commented on text/reporting_assurance.tex:
```diff
@@ -12,18 +12,20 @@ \section{Reporting and Assurance}
 
 \subsection{State}
-The state of the reporting and availability portion of the protocol is largely contained within $\reports$, which tracks the work-reports which have been reported but are not yet known to be available to a super-majority of validators, together with the time at which each was reported. As mentioned earlier, only one report may be assigned to a core at any given time. Formally:
+The state of the reporting and availability portion of the protocol is largely contained within $\availassignments$, which tracks the work-report guarantees which have been reported but are not yet known to be available to a super-majority of validators, together with the time at which each was reported. As mentioned earlier, only one guarantee may be assigned to a core at any given time. Formall ...
```
This is where I think we should name the term "availability assignment" somehow. However, it is not clear to me why this wording was chosen. Is it because we assign each guarantee to a core, while it is unavailable? And iiuc an entry stops being in \availassignments once it becomes available?

zdave-parity replied:
This now uses the term "availability assignment".

davxy reviewed (dismissed):
Looks good

zdave-parity committed:
Attempt to improve wording and terminology in various places

In particular, try to use "availability assignment" everywhere rather
than eg "pending report" or "pending guarantee".

zdave-parity commented on text/auditing.tex:
```diff
@@ -29,7 +29,7 @@ \subsection{Data Fetching}
 \label{sec:auditselection}
 
-Each validator shall perform auditing duties on each valid block received. Since we are entering off-chain logic, and we cannot assume consensus, we henceforth consider ourselves a specific validator of index $v$ and assume ourselves focused on some recent block $\block$ with other terms corresponding to the state-transition implied by that block, so $\reports$ is said block's prior core-allocati ...
+Each validator shall perform auditing duties on each valid block received. Since we are entering off-chain logic, and we cannot assume consensus, we henceforth consider ourselves a specific validator of index $v$ and assume ourselves focused on some recent block $\block$ with other terms corresponding to the state-transition implied by that block, so $\availassignments$ is said block's prior avai ...
```
Hopefully better now.

zdave-parity commented on text/auditing.tex:
```diff
@@ -176,7 +176,7 @@ \subsection{Selection of Reports}
 Furthermore, we pointedly disregard chains which include the accumulation of a report which we know at least $\onethird$ of validators judge as being invalid. Any chains including such a block are not eligible for authoring on. The \emph{best block}, \ie that on which we build new blocks, is defined as the chain with the most regular Safrole blocks which does \emph{not} contain any such disregard ...
 
-As a block author, we include a judgment extrinsic which collects judgment signatures together and reports them on-chain. In the case of a non-valid judgment (\ie one which is not two-thirds-plus-one of judgments confirming validity) then this extrinsic will be introduced in a block in which accumulation of the non-valid work-report is about to take place. The non-valid judgment extrinsic removes ...
+As a block author, we include a disputes extrinsic which collects judgment signatures together into verdicts and reports them on-chain. In the case of a non-valid verdict (\ie one which is not two-thirds-plus-one of judgments confirming validity) then this verdict will be introduced in a block in which accumulation of the non-valid work-report is about to take place. The non-valid verdict in the  ...
```
Reworded this. Hopefully clearer now.

alxmirap reviewed (commented):
LGTM

zdave-parity committed:
Support smaller validator sets

The designate host call now has a length argument, permitting validator
sets smaller than 1023 validators to be specified. For simplicity, we
require that the size is a multiple of 3 and at least 6 (i.e. 2 cores).

The number of cores remains fixed at 341, but the number of "active"
cores scales with the number of active validators. It is not possible to
report work-packages on inactive cores.

When there are fewer validators, we permit each validator to generate
more Safrole tickets, to ensure the accumulator can be filled.

For simplicity, the erasure coding rate and piece size are defined such
that segments are always a whole number of pieces and do not require
padding. In practice this means that the erasure coding rate is
suboptimal for many validator set sizes; it is recommended that the
validator set size is picked from the set of optimal sizes given in the
erasure coding section.

gavofyork reviewed (changes requested):
Needs rebasing...

zdave-parity commented:
Rebased.

Also got rid of `\accinput`. I hadn't realised this was also using `\mathbb{I}` as it wasn't listed in the notation section. It's only really needed in one place (definition of the accumulation invocation function); it's simpler to just write `U cup X` there instead.

This will likely need rebasing again on top of #514, or the other way around. I'll do that on a separate branch for now...

zdave-parity committed:
Merge branch 'var-vals' into guarantees-in-state

zdave-parity commented:
> This will likely need rebasing again on top of #514, or the other way around. I'll do that on a separate branch for now...

Branch merging this with #514 available [here](/gavofyork/graypaper/tree/dave/guarantees-in-state-plus-var-vals).

gavofyork commented:
@zdave-parity what's the preferred way forward here? Merge #514 first and then rebase this to `main` and review? Or merge the new branch into #514 and then review/merge them both in one go?

zdave-parity commented:
> @zdave-parity what's the preferred way forward here? Merge #514 first and then rebase this to `main` and review? Or merge the new branch into #514 and then review/merge them both in one go?

Probably best to merge one at a time. Either way round works, I've done the merge already so should be trivial to update either PR after the other goes in.

zdave-parity committed:
Merge remote-tracking branch 'origin/main' into HEAD
